compiler/rustc_middle/src/ty/sty.rs

```diff
@@ -2372,6 +2372,46 @@ impl<'tcx> Ty<'tcx> {
             }
         }
     }
+
+    /// Fast path helper for primitives which are always `Copy` and which
+    /// have a side-effect-free `Clone` impl.
+    ///
+    /// Returning true means the type is known to be pure and `Copy+Clone`.
+    /// Returning `false` means nothing -- could be `Copy`, might not be.
+    ///
+    /// This is mostly useful for optimizations, as there are the types
+    /// on which we can replace cloning with dereferencing.
+    pub fn is_trivially_pure_clone_copy(self) -> bool {
+        match self.kind() {
+            ty::Bool | ty::Char | ty::Never => true,
+
+            // These aren't even `Clone`
+            ty::Str | ty::Slice(..) | ty::Foreign(..) | ty::Dynamic(..) => false,
+
+            ty::Int(..) | ty::Uint(..) | ty::Float(..) => true,
+
+            // Both the voldemort ZSTs and the pointers are fine.
+            ty::FnDef(..) | ty::FnPtr(..) => true,
```

Comment:
Well, I added that check because a 100-tuple isn't "trivially" anything, so 12 is as good a stopping point as any, but note that *for `Copy`* this isn't an issue, apparently: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=a2d7a0a7f7e028f7ef5d9910b8acc684